File: backend/utils/refund_allocator.py
"""
refund_allocator — Payment-aware refund routing for invoice corrections and returns.

Given an invoice and a refund amount, compute how the refund should be allocated
across three buckets in priority order:

    1. AR REDUCTION    — applied to the still-open balance (invoice.balance).
                         No money leaves any wallet; just reduces what the
                         customer owes.
    2. DIGITAL REVERSAL — reverses paid digital payments (GCash, Maya, Bank
                         Transfer, etc.), debiting the digital wallet by the
                         matching channel + ref number.
    3. CASH REFUND     — debits the cashier wallet for the residual amount.

This intentionally mirrors how `void_invoice` reverses cashflow in
`routes/invoices.py` (lines ~1240-1280) so corrections behave consistently
with voids.

Why this ordering?
- AR-first is safe: it only shrinks an obligation, never moves real money.
- Digital-then-cash matches the physical reality: digital payments leave a
  refundable trail (txn ID, sender), so reversing the same channel is most
  auditable. Cash is the catch-all when nothing else can absorb the refund.
"""
from typing import List, Dict, Any
from .helpers import is_digital_payment


def _safe_float(v, default=0.0) -> float:
    try:
        return float(v)
    except (TypeError, ValueError):
        return default


def summarize_invoice_payments(invoice: Dict[str, Any]) -> Dict[str, Any]:
    """
    Walk an invoice and produce a per-channel payment summary.

    Returns:
      {
        "grand_total":   float,
        "open_balance":  float,   # still owed (== invoice.balance)
        "cash_paid":     float,
        "digital_paid":  float,
        "check_paid":    float,
        # Detailed payment rows, newest-first ordering preserved for reversal:
        "payment_rows": [
            {"id", "amount", "method", "fund_source", "kind": "cash"|"digital"|"check",
             "platform", "ref_number", "source": "initial"|"subsequent"}
        ]
      }

    `payment_rows` is the canonical source of truth for the allocator. It
    composes two inputs:
      a) The INITIAL payment captured at invoice creation. `amount_paid` is
         carried at top-level along with `payment_method`, optional
         `cash_amount`/`digital_amount`/`digital_platform` for splits.
      b) The SUBSEQUENT `payments[]` array — AR payments recorded later via
         `record_invoice_payment`.

    We build the rows in chronological order (initial first, then payments[]
    in their natural order). The allocator will reverse this for "newest-first
    reversal".
    """
    grand_total = _safe_float(invoice.get("grand_total"))
    balance = _safe_float(invoice.get("balance"))
    payment_type = (invoice.get("payment_type") or "").lower()
    payment_method = invoice.get("payment_method") or "Cash"

    rows: List[Dict[str, Any]] = []

    # ── Initial payment row(s) reconstructed from top-level fields ─────────
    initial_amount = _safe_float(invoice.get("amount_paid")) - _sum_subsequent(invoice)
    initial_amount = max(0.0, round(initial_amount, 2))

    if initial_amount > 0:
        if payment_type == "split":
            cash_part = _safe_float(invoice.get("cash_amount"))
            digital_part = _safe_float(invoice.get("digital_amount"))
            # Fall back proportionally if split fields are missing
            if cash_part <= 0 and digital_part <= 0:
                cash_part = initial_amount / 2
                digital_part = initial_amount / 2
            if cash_part > 0:
                rows.append(_mk_row(initial_amount=cash_part, method="Cash",
                                    fund_source="cashier", source="initial-split-cash"))
            if digital_part > 0:
                rows.append(_mk_row(
                    initial_amount=digital_part,
                    method=invoice.get("digital_platform") or "Digital",
                    fund_source="digital",
                    platform=invoice.get("digital_platform") or "Digital",
                    ref_number=invoice.get("digital_ref_number", ""),
                    source="initial-split-digital",
                ))
        else:
            fund_source = "digital" if is_digital_payment(payment_method) else "cashier"
            rows.append(_mk_row(
                initial_amount=initial_amount,
                method=payment_method,
                fund_source=fund_source,
                platform=invoice.get("digital_platform", payment_method) if fund_source == "digital" else "",
                ref_number=invoice.get("digital_ref_number", "") if fund_source == "digital" else "",
                source="initial",
            ))

    # ── Subsequent payments (AR payments) ──────────────────────────────────
    for pmt in invoice.get("payments", []) or []:
        if pmt.get("voided"):
            continue
        amt = _safe_float(pmt.get("amount"))
        if amt <= 0:
            continue
        method = pmt.get("method", "Cash")
        fund_source = pmt.get("fund_source") or (
            "digital" if is_digital_payment(method) else "cashier"
        )
        rows.append(_mk_row(
            initial_amount=amt,
            method=method,
            fund_source=fund_source,
            platform=pmt.get("digital_platform", method) if fund_source == "digital" else "",
            ref_number=pmt.get("digital_ref_number", pmt.get("reference", "")) if fund_source == "digital" else "",
            payment_id=pmt.get("id", ""),
            source="subsequent",
        ))

    cash_paid = round(sum(r["amount"] for r in rows if r["kind"] == "cash"), 2)
    digital_paid = round(sum(r["amount"] for r in rows if r["kind"] == "digital"), 2)
    check_paid = round(sum(r["amount"] for r in rows if r["kind"] == "check"), 2)

    return {
        "grand_total":  round(grand_total, 2),
        "open_balance": round(balance, 2),
        "cash_paid":    cash_paid,
        "digital_paid": digital_paid,
        "check_paid":   check_paid,
        "payment_rows": rows,
    }


def _sum_subsequent(invoice: Dict[str, Any]) -> float:
    total = 0.0
    for pmt in invoice.get("payments", []) or []:
        if pmt.get("voided"):
            continue
        total += _safe_float(pmt.get("amount"))
    return round(total, 2)


def _classify(method: str, fund_source: str) -> str:
    m = (method or "").lower().strip()
    fs = (fund_source or "").lower().strip()
    if m in ("check", "cheque"):
        return "check"
    if fs == "digital" or is_digital_payment(method):
        return "digital"
    return "cash"


def _mk_row(initial_amount: float, method: str, fund_source: str,
            platform: str = "", ref_number: str = "",
            payment_id: str = "", source: str = "") -> Dict[str, Any]:
    return {
        "id":           payment_id,
        "amount":       round(_safe_float(initial_amount), 2),
        "method":       method,
        "fund_source":  fund_source,
        "platform":     platform or "",
        "ref_number":   ref_number or "",
        "source":       source,
        "kind":         _classify(method, fund_source),
    }


def compute_refund_allocation(invoice: Dict[str, Any],
                              refund_amount: float) -> Dict[str, Any]:
    """
    Allocate `refund_amount` across AR/digital/cash buckets for `invoice`.

    Returns:
      {
        "total":                  float,
        "ar_reduction":           float,
        "digital_refunds":        [ {amount, method, platform, ref_number, payment_id} ],
        "check_refunds":          [ {amount, ref_number, payment_id} ],
        "cash_refund":            float,
        "remaining_unallocated":  float,   # 0 unless refund > grand_total
        "summary":                summarize_invoice_payments(invoice),
      }

    Rules:
    - AR-first: never exceed `invoice.balance`.
    - Then walk digital payment rows newest-first, taking up to each row's
      remaining amount.
    - Then cash from the cashier wallet for the residual.
    - Check payments are NOT auto-refunded (we surface them in `check_refunds`
      so the caller can record manual handling).
    """
    refund_amount = max(0.0, round(_safe_float(refund_amount), 2))
    summary = summarize_invoice_payments(invoice)

    remaining = refund_amount

    # 1. AR
    ar_take = min(remaining, summary["open_balance"])
    remaining = round(remaining - ar_take, 2)

    # 2. Digital — newest first
    digital_refunds: List[Dict[str, Any]] = []
    if remaining > 0 and summary["digital_paid"] > 0:
        digital_rows = [r for r in summary["payment_rows"] if r["kind"] == "digital"]
        for r in reversed(digital_rows):
            if remaining <= 0:
                break
            take = min(remaining, r["amount"])
            if take <= 0:
                continue
            digital_refunds.append({
                "amount":      round(take, 2),
                "method":      r["method"],
                "platform":    r["platform"],
                "ref_number":  r["ref_number"],
                "payment_id":  r["id"],
            })
            remaining = round(remaining - take, 2)

    # 3. Check — surface but DO NOT auto-refund. Cashier must reissue/cancel
    #    the check manually. Caller decides whether to push residual into
    #    cash or leave it unallocated.
    check_refunds: List[Dict[str, Any]] = []
    if remaining > 0 and summary["check_paid"] > 0:
        check_rows = [r for r in summary["payment_rows"] if r["kind"] == "check"]
        for r in reversed(check_rows):
            if remaining <= 0:
                break
            take = min(remaining, r["amount"])
            if take <= 0:
                continue
            check_refunds.append({
                "amount":     round(take, 2),
                "ref_number": r["ref_number"],
                "payment_id": r["id"],
            })
            # Check refunds DON'T reduce `remaining` automatically: the
            # caller must decide whether to fall through to cash or leave
            # it as a manual task. Default behaviour: surface and fall
            # through.

    # 4. Cash residual — bounded by what was actually paid in cash (so we
    #    never debit more than the cashier originally received). Anything
    #    above that bound surfaces as `remaining_unallocated` for the caller
    #    to handle (typically: reject, or escalate to admin).
    cash_residual_requested = max(0.0, remaining)
    cash_refund = round(min(cash_residual_requested, summary["cash_paid"]), 2)

    # Compute final unallocated (== 0 in healthy flows). Anything that
    # remains here means refund > grand_total or some data corruption.
    allocated = (
        ar_take
        + sum(d["amount"] for d in digital_refunds)
        + cash_refund
    )
    unallocated = round(max(0.0, refund_amount - allocated), 2)

    return {
        "total":                   round(refund_amount, 2),
        "ar_reduction":            round(ar_take, 2),
        "digital_refunds":         digital_refunds,
        "check_refunds":           check_refunds,
        "cash_refund":             cash_refund,
        "remaining_unallocated":   unallocated,
        "summary":                 summary,
    }

File: frontend/src/lib/refundAllocator.js
/**
 * refundAllocator — frontend mirror of backend's `compute_refund_allocation`.
 *
 * Used by terminal modals (Return & Refund, Update for Incomplete Stock) to
 * preview how a refund will be routed before the cashier confirms:
 *
 *    1. AR FIRST   → shrinks invoice.balance, no money moves
 *    2. DIGITAL    → reverses digital wallet by paid channel/ref number
 *    3. CASH       → debits cashier wallet for the residual
 *
 * Stays intentionally pure — no fetches, no state. Caller hands it an
 * invoice-shape object and a refund amount.
 */

const DIGITAL_METHODS = new Set([
  'gcash', 'maya', 'paymaya', 'bank transfer', 'instapay', 'pesonet',
  'paypal', 'shopee pay', 'grabpay', 'coins.ph', 'seabank',
  'digital', 'e-wallet', 'online transfer', 'mobile payment',
]);

function isDigital(method) {
  if (!method) return false;
  const m = String(method).toLowerCase().trim();
  if (['cash', 'check', 'cheque', 'credit', 'partial', 'split', ''].includes(m)) return false;
  return DIGITAL_METHODS.has(m) || true;
}

function classify(method, fundSource) {
  const m = (method || '').toLowerCase().trim();
  const fs = (fundSource || '').toLowerCase().trim();
  if (m === 'check' || m === 'cheque') return 'check';
  if (fs === 'digital' || isDigital(method)) return 'digital';
  return 'cash';
}

function num(v) { const f = parseFloat(v); return isNaN(f) ? 0 : f; }

export function summarizeInvoicePayments(invoice) {
  const grandTotal = num(invoice?.grand_total);
  const balance = num(invoice?.balance);
  const paymentType = (invoice?.payment_type || '').toLowerCase();
  const paymentMethod = invoice?.payment_method || 'Cash';

  const rows = [];

  // Subtract subsequent payments to derive the initial payment amount.
  const subsequentTotal = (invoice?.payments || [])
    .filter(p => !p?.voided)
    .reduce((s, p) => s + num(p?.amount), 0);
  const initialAmount = Math.max(0, +(num(invoice?.amount_paid) - subsequentTotal).toFixed(2));

  if (initialAmount > 0) {
    if (paymentType === 'split') {
      let cashPart = num(invoice?.cash_amount);
      let digitalPart = num(invoice?.digital_amount);
      if (cashPart <= 0 && digitalPart <= 0) {
        cashPart = initialAmount / 2;
        digitalPart = initialAmount / 2;
      }
      if (cashPart > 0) {
        rows.push({ amount: +cashPart.toFixed(2), method: 'Cash',
                    fund_source: 'cashier', kind: 'cash', source: 'initial-split-cash' });
      }
      if (digitalPart > 0) {
        rows.push({
          amount: +digitalPart.toFixed(2),
          method: invoice?.digital_platform || 'Digital',
          fund_source: 'digital', kind: 'digital',
          platform: invoice?.digital_platform || 'Digital',
          ref_number: invoice?.digital_ref_number || '',
          source: 'initial-split-digital',
        });
      }
    } else {
      const fs = isDigital(paymentMethod) ? 'digital' : 'cashier';
      rows.push({
        amount: +initialAmount.toFixed(2),
        method: paymentMethod,
        fund_source: fs,
        kind: fs === 'digital' ? 'digital' : 'cash',
        platform: fs === 'digital' ? (invoice?.digital_platform || paymentMethod) : '',
        ref_number: fs === 'digital' ? (invoice?.digital_ref_number || '') : '',
        source: 'initial',
      });
    }
  }

  for (const pmt of (invoice?.payments || [])) {
    if (pmt?.voided) continue;
    const amt = num(pmt?.amount);
    if (amt <= 0) continue;
    const method = pmt?.method || 'Cash';
    const fs = pmt?.fund_source || (isDigital(method) ? 'digital' : 'cashier');
    rows.push({
      id: pmt.id || '',
      amount: +amt.toFixed(2),
      method, fund_source: fs,
      kind: classify(method, fs),
      platform: fs === 'digital' ? (pmt.digital_platform || method) : '',
      ref_number: fs === 'digital' ? (pmt.digital_ref_number || pmt.reference || '') : '',
      source: 'subsequent',
    });
  }

  const cashPaid = +rows.filter(r => r.kind === 'cash').reduce((s, r) => s + r.amount, 0).toFixed(2);
  const digitalPaid = +rows.filter(r => r.kind === 'digital').reduce((s, r) => s + r.amount, 0).toFixed(2);
  const checkPaid = +rows.filter(r => r.kind === 'check').reduce((s, r) => s + r.amount, 0).toFixed(2);

  return {
    grand_total: +grandTotal.toFixed(2),
    open_balance: +balance.toFixed(2),
    cash_paid: cashPaid,
    digital_paid: digitalPaid,
    check_paid: checkPaid,
    payment_rows: rows,
  };
}

export function computeRefundAllocation(invoice, refundAmount) {
  const refund = Math.max(0, num(refundAmount));
  const summary = summarizeInvoicePayments(invoice);
  let remaining = refund;

  const arTake = Math.min(remaining, summary.open_balance);
  remaining = +(remaining - arTake).toFixed(2);

  const digitalRefunds = [];
  if (remaining > 0 && summary.digital_paid > 0) {
    const digitalRows = summary.payment_rows.filter(r => r.kind === 'digital');
    for (let i = digitalRows.length - 1; i >= 0; i--) {
      if (remaining <= 0) break;
      const r = digitalRows[i];
      const take = Math.min(remaining, r.amount);
      if (take <= 0) continue;
      digitalRefunds.push({
        amount: +take.toFixed(2),
        method: r.method,
        platform: r.platform || '',
        ref_number: r.ref_number || '',
        payment_id: r.id || '',
      });
      remaining = +(remaining - take).toFixed(2);
    }
  }

  // Cash bounded by what was actually paid in cash.
  const cashResidual = Math.max(0, remaining);
  const cashRefund = +Math.min(cashResidual, summary.cash_paid).toFixed(2);

  const allocated = +(arTake
    + digitalRefunds.reduce((s, d) => s + d.amount, 0)
    + cashRefund).toFixed(2);
  const unalloc = +Math.max(0, refund - allocated).toFixed(2);

  return {
    total: +refund.toFixed(2),
    ar_reduction: +arTake.toFixed(2),
    digital_refunds: digitalRefunds,
    check_refunds: [],
    cash_refund: cashRefund,
    remaining_unallocated: unalloc,
    summary,
  };
}
